roadmap/implementers-guide/src/runtime/ump.md

```diff
@@ -0,0 +1,98 @@
+# UMP Module
+
+## Storage
+
+General storage entries
+
+```rust
+/// Paras that are to be cleaned up at the end of the session.
+/// The entries are sorted ascending by the para id.
+OutgoingParas: Vec<ParaId>;
+```
+
+Storage related to UMP
+
+```rust
+/// The messages waiting to be handled by the relay-chain originating from a certain parachain.
+///
+/// Note that some upward messages might have been already processed by the inclusion logic. E.g.
+/// channel management messages.
+///
+/// The messages are processed in FIFO order.
+RelayDispatchQueues: map ParaId => Vec<UpwardMessage>;
+/// Size of the dispatch queues. Caches sizes of the queues in `RelayDispatchQueue`.
+///
+/// First item in the tuple is the count of messages and second
+/// is the total length (in bytes) of the message payloads.
+///
+/// Note that this is an auxilary mapping: it's possible to tell the byte size and the number of
+/// messages only looking at `RelayDispatchQueues`. This mapping is separate to avoid the cost of
+/// loading the whole message queue if only the total size and count are required.
+///
+/// Invariant:
+/// - The set of keys should exactly match the set of keys of `RelayDispatchQueues`.
+RelayDispatchQueueSize: map ParaId => (u32, u32); // (num_messages, total_bytes)
+/// The ordered list of `ParaId`s that have a `RelayDispatchQueue` entry.
+///
+/// Invariant:
+/// - The set of items from this vector should be exactly the set of the keys in
+///   `RelayDispatchQueues` and `RelayDispatchQueueSize`.
+NeedsDispatch: Vec<ParaId>;
+/// This is the para that gets dispatched first during the next upward dispatchable queue
+/// execution round.
+///
+/// Invariant:
+/// - If `Some(para)`, then `para` must be present in `NeedsDispatch`.
+NextDispatchRoundStartWith: Option<ParaId>;
+```
+
+
+## Initialization
+
+No initialization routine runs for this module.
+
+## Routines
+
+Candidate Acceptance Function:
+
+* `check_upward_messages(P: ParaId, Vec<UpwardMessage>`):
+    1. Checks that there are at most `config.max_upward_message_num_per_candidate` messages.
+    1. Checks that no message exceeds `config.max_upward_message_size`.
+    1. Verify that `RelayDispatchQueueSize` for `P` has enough capacity for the messages
+
+Candidate Enactment:
+
+* `enact_upward_messages(P: ParaId, Vec<UpwardMessage>)`:
+    1. Process each upward message `M` in order:
+        1. Append the message to `RelayDispatchQueues` for `P`
+        1. Increment the size and the count in `RelayDispatchQueueSize` for `P`.
+        1. Ensure that `P` is present in `NeedsDispatch`.
+
+The following routine is intended to be called in the same time when `Paras::schedule_para_cleanup` is called.
+
+`schedule_para_cleanup(ParaId)`:
```

Comment:
`para` should be named.

Reply:
again, the same logic as [here](https://github.com/paritytech/polkadot/pull/1939#discussion_r523009204)